### PR TITLE
Use PSR-7 responses in additional actions

### PR DESCRIPTION
**Pull request by TimWolla**
Blocked on #4436

### CODE REVIEW

File: wcfsetup/install/files/lib/util/HeaderUtil.class.php [hunk 1]
```diff
@@ -106,6 +107,28 @@ public static function sendNoCacheHeaders()
         @\header('Cache-Control: max-age=0, no-cache, no-store, must-revalidate');
     }
 
+    /**
+     * Returns an response matching the given response with headers preventing in-browser
+     * caching attached.
+     */
+    public static function withNoCacheHeaders(ResponseInterface $response): ResponseInterface
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/util/HeaderUtil.class.php:114]: I wonder if this should be moved somewhere to `lib/system/io/http/`?

File: wcfsetup/install/files/lib/util/HeaderUtil.class.php [hunk 2]
```diff
@@ -106,6 +107,28 @@ public static function sendNoCacheHeaders()
         @\header('Cache-Control: max-age=0, no-cache, no-store, must-revalidate');
     }
 
+    /**
+     * Returns an response matching the given response with headers preventing in-browser
+     * caching attached.
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/util/HeaderUtil.class.php:112]: `@since`

File: wcfsetup/install/files/lib/action/AJAXUploadAction.class.php [hunk 3]
```diff
@@ -34,10 +35,10 @@ protected function sendResponse()
             parent::sendResponse();
         }
 
+        $response = new JsonResponse($this->response);
         // IE9 is mad if iframe response is application/json
-        \header('Content-type: text/plain');
-        echo JSON::encode($this->response);
+        $response = $response->withHeader('content-type', 'text/plain');
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/action/AJAXUploadAction.class.php:40]: Please just remove this non-sense, it is an old legacy hack for IE9 :D

File: wcfsetup/install/files/lib/action/FacebookAuthAction.class.php [hunk 4]
```diff
@@ -154,9 +154,9 @@ protected function processUser(OauthUser $oauthUser)
 
                 WCF::getSession()->register('__oauthUser', $oauthUser);
 
-                HeaderUtil::redirect(LinkHandler::getInstance()->getLink('AccountManagement') . '#3rdParty');
-
-                exit;
+                return new RedirectResponse(
+                    LinkHandler::getInstance()->getLink('AccountManagement') . '#3rdParty'
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/action/FacebookAuthAction.class.php:158]: Can you replace this one with `getControllerLink()`?

File: wcfsetup/install/files/lib/action/GithubAuthAction.class.php [hunk 5]
```diff
@@ -144,9 +144,9 @@ protected function processUser(OauthUser $oauthUser)
 
                 WCF::getSession()->register('__oauthUser', $oauthUser);
 
-                HeaderUtil::redirect(LinkHandler::getInstance()->getLink('AccountManagement') . '#3rdParty');
-
-                exit;
+                return new RedirectResponse(
+                    LinkHandler::getInstance()->getLink('AccountManagement') . '#3rdParty'
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/action/GithubAuthAction.class.php:148]: Can you replace this one with `getControllerLink()`?

File: wcfsetup/install/files/lib/action/GoogleAuthAction.class.php [hunk 6]
```diff
@@ -166,9 +166,9 @@ protected function processUser(OauthUser $oauthUser)
 
                 WCF::getSession()->register('__oauthUser', $oauthUser);
 
-                HeaderUtil::redirect(LinkHandler::getInstance()->getLink('AccountManagement') . '#3rdParty');
-
-                exit;
+                return new RedirectResponse(
+                    LinkHandler::getInstance()->getLink('AccountManagement') . '#3rdParty'
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/action/GoogleAuthAction.class.php:170]: Can you replace this one with `getControllerLink()`?

File: wcfsetup/install/files/lib/action/MessageQuoteAction.class.php [hunk 7]
```diff
@@ -131,12 +132,10 @@ public function execute()
         WCF::getSession()->disableUpdate();
 
         if ($returnValues !== null) {
-            // send JSON-encoded response
-            \header('Content-type: application/json; charset=UTF-8');
-            echo JSON::encode($returnValues);
+            return new JsonResponse($returnValues);
+        } else {
+            return new EmptyResponse(200);
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/action/MessageQuoteAction.class.php:137]: You can get rid of the `else` here.

File: wcfsetup/install/files/lib/util/HeaderUtil.class.php (hunk 1 above)
Comment (dtdesign) [reply wcfsetup/install/files/lib/util/HeaderUtil.class.php:116]: This would only make sense if you're looking forward to phase out this class?

**Review by dtdesign — CHANGES_REQUESTED** [review]
LGTM

File: wcfsetup/install/files/lib/action/AJAXUploadAction.class.php (hunk 3 above)
Comment (TimWolla) [reply wcfsetup/install/files/lib/action/AJAXUploadAction.class.php:40]: As explained in [the commit message](https://github.com/WoltLab/WCF/pull/4447/commits/4dfe137efc1fb1dfea1ae435335749218d13c2cb):

> The IE 9 fallback technically should no longer be required, but the (legacy)
> JavaScript still references the `isFallback` parameter.

I'd rather clean this up once the JavaScript is cleaned up as well to not mix 2 unrelated things.

File: wcfsetup/install/files/lib/util/HeaderUtil.class.php (hunk 1 above)
Comment (TimWolla) [reply wcfsetup/install/files/lib/util/HeaderUtil.class.php:116]: Looking at the methods in `HeaderUtil`:

- `setCookie`: This is one of the issues that should be solved with PSR-7. Will need to be adapted, might be replaced by: https://github.com/dflydev/dflydev-fig-cookies
- `getCookieDomain`: Might be simplified once multi-domain setups are gone?
- `sendHeaders`: Will become obsolete with PSR-7 and is then better solved using a middleware.
- `sendNoCacheHeaders`: This is replaced by the method we are currently discussing.
- `parseOutput`: Is intimately tied to `sendHeaders`.
- `redirect`: Will be replaced by `RedirectResponse`.
- `delayedRedirect`: Should probably be replaced by a specific `DelayedRedirectResponse`.

Basically all of the methods in `HeaderUtil` will become obsolete with the migration to PSR-7. Thus why I want to discuss whether re-architecting this might make sense. For Guzzle we already have the RedirectGuard as a single-purpose class:

https://github.com/WoltLab/WCF/blob/master/wcfsetup/install/files/lib/system/io/http/RedirectGuard.class.php

File: wcfsetup/install/files/lib/action/MessageQuoteAction.class.php (hunk 7 above)
Comment (TimWolla) [reply wcfsetup/install/files/lib/action/MessageQuoteAction.class.php:137]: In this case it's intentional to make it more clear that these are the only 2 cases. An early return IMO would not be as fitting.

The `EmptyResponse` case can also be removed once the `markForRemoval` action is removed (I deprecated it in a previous PR).

File: wcfsetup/install/files/lib/util/HeaderUtil.class.php (hunk 1 above)
Comment (dtdesign) [reply wcfsetup/install/files/lib/util/HeaderUtil.class.php:116]: I agree, spreading things out while still having a mix of everything certainly does not improve the situation.

File: wcfsetup/install/files/lib/action/AJAXUploadAction.class.php (hunk 3 above)
Comment (dtdesign) [reply wcfsetup/install/files/lib/action/AJAXUploadAction.class.php:40]: OK, I was under the false impression that the fallback was removed a while ago, because it is effectively unsupported.

**Review by dtdesign — APPROVED** [review]
